functorch/csrc/BatchRulesScatterOps.cpp

```diff
@@ -111,6 +111,45 @@ Tensor index_plumbing(const Tensor & self, const List<optional<Tensor>> & indice
   return makeBatched(std::get<0>(results), std::get<1>(results), cur_level);
 }
 
+namespace {
+  std::tuple<Tensor, std::vector<optional<Tensor>>, Tensor>
+  index_put_batch_rule_helper(const Tensor &self,
+                              optional<int64_t> self_bdim,
+                              ArrayRef<optional<Tensor>> indices,
+                              ArrayRef<optional<int64_t>> indices_bdims,
+                              const Tensor &values,
+                              optional<int64_t> values_bdim) {
+
+    auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+    auto values_logical_rank = rankWithoutBatchDim(values, values_bdim);
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor values_ = moveBatchDimToFront(values, values_bdim);
+    TORCH_INTERNAL_ASSERT(indices.size() == indices_bdims.size());
+    std::vector<optional<Tensor>> indices_ = batchIndices(indices, indices_bdims, self_.size(0), self_bdim, values_bdim);
+
+    // handle case when prefix unit dimensions are stripped in setitem.
+    // values of shape `[1, 1, 2]` becomes `[2]`
+    // This leads to problem if values was actually batched.
```

Comment:
Okay I think I understand what the problem is. It doesn't seem like a problem specific to setitem -- you can also run into this by directly calling index_put_. Please let me know if this is correct:

The problem is that `tensor.index_put_(idxs, value)'s batching rule didn't handle broadcasting the value. If we have ```
tensor =  torch.zeros(3, 4)
value = torch.ones(2)
idxs = (torch.tensor([0]), torch.tensor([1, 2]))
```
and we do `tensor.index_put_(idxs, value)`, tensor[idxs] is a tensor of shape (1, 2) and value is a tensor of shape (2,). value broadcasts up to the size of tensor[idxs].

However, if the values is batched, we have to be very careful: if the value is a BatchedTensor(Tensor[B, 2]), then we must insert additional 1's before invoking `index_put_` on regular tensors so that the value broadcasts up to the size of tensor[idxs].

The way we're determining the (logical) rank of tensor[idxs] is that it is equal to the (logical) rank of tensor, therefore the solution is to put 1's into the value until it has the same (logical) rank as that of tensor.

Reply:
If this sounds correct:
- we should rewrite the comment string here to make it clearer that this is what is going on
- we should add some explicit tests for index_put_ under vmap

Reply:
Ah. Right. That is indeed the case. Sorry I was confused. Have updated the code and added a two test cases.

Thanks :)